src/core/server/ui_settings/ui_settings_client.ts

```diff
@@ -147,17 +143,50 @@ export class UiSettingsClient implements IUiSettingsClient {
     return defaultsDeep(userProvided, this.defaults);
   }
 
+  private validateKey(key: string, value: unknown) {
+    const definition = this.defaults[key];
+    if (value === null || definition === undefined) return;
+    if (definition.schema) {
+      definition.schema.validate(value, {}, `validation [${key}]`);
+    }
+  }
+
+  private onWriteHook(changes: Record<string, unknown>) {
+    for (const key of Object.keys(changes)) {
+      this.assertUpdateAllowed(key);
+    }
+
+    for (const [key, value] of Object.entries(changes)) {
+      this.validateKey(key, value);
+    }
+  }
+
+  private onReadHook<T = unknown>(values: Record<string, unknown>) {
+    // write the userValue for each key stored in the saved object that is not overridden
+    // validate value read from saved objects as it can be changed via SO API
+    const filteredValues: UserProvided<T> = {};
+    for (const [key, userValue] of Object.entries(values)) {
+      if (userValue === null || this.isOverridden(key)) continue;
```

Comment:
I think this might be simpler & possibly safer if we validated the fully resolved settings (user provided + defaults) rather than just the user provided values. There's a small performance penalty for that, but I imagine it's negligible?

I just worry about this getting refactored at some point and the validation passing on settings that may not actually match the expected types 100%.

Reply:
We will validate defaults and overrides in the start phase.
```
As a bonus, uiSettings service validates on start defaults values. doesn't work currently since all declarations are done in LP
```
I don't wait to waste CPU cycles since users usually adjust a handful of settings, so I'd rather prioritize https://github.com/elastic/kibana/issues/59755 to move definitions to the NP.